projects/evm-compatible-bridge/evm_contract_design.md

```diff
@@ -0,0 +1,141 @@
+
+The bridge contract will also be required to have some user and admin methods.
+
+#### User methods
+
+- `getTokenId`
+  - All tokens supported by the bridge will have a unique id, this includes tokens native to the EVM chain and tokens managed by the contract.
+  - The user will provide a token address and optionally an integer which will be the ERC-1155 token id (for tokens managed by the bridge).
+  - For the chain native token, the token id will always be `0` so this call is unnecessary.
+- `crossTokens`
+  - This method will be called by the user to send tokens to the bridge.
+  - The bridge contract will check if the token is supported and if it is not, it will fail the transaction.
+  - The user will send tokens to the bridge contract address.
+  - The bridge will register this request with the federation contract.
+  - The bridge contract will emit an event to log the request.
+  - The amount of tokens must be valid according to the granularity rules.
+- `requestSupportForToken`
+  - The user will call this method to request support for a token (native to the EVM chain).
+  - This method will register a unique id and register this request with the federation contract.
+  - The bridge contract will emit an event to log the request.
+- All features supoprted by the ERC-1155 token standard.
+
+#### Admin methods
+
+This contract is "ownable" and the admin will be able to make some operations, for instance:
+
+- Contract suspension (i.e. pausable contract)
+  - In case of a security breach or legal obstruction we may have to halt operations.
+  - Example implementation [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/Pausable.sol)
+- Burn and mint tokens
+  - The admin should be able to manage tokens in the bridge.
+  - The admin can use this to refund tokens or make manual corrections.
+- Redeem fees
+  - The admin can redeem fees collected by the bridge from the transactions.
+- Change federation contract
+  - The admin can change the federation contract address.
+- Set token id
+  - The admin should be able to correct and change the mapping of equivalent tokens.
+  - This is to correct any mistakes or to manually support new tokens.
+
+#### Bridge contract storage
+
+##### _Token equivalency_
+
+Other than the storage required to implement the ERC-1155 token standard we require additional storage to track which tokens are allowed in the bridge and to map the equivalency of Hathor and EVM tokens.
+
+For tokens native to the EVM compatible chain we require an identification called `nativeTokenId` which we will map to a local id.
+This local id will be mapped to a Hathor token uid.
+The same will be done in reverse, a Hathor native token when crossing to the EVM chain will mint an equivalent token in our ERC-1155 contract.
+A token created in this contract will have a `uint256` id.
+
+```solidity
+struct nativeTokenId {
+  address contract;
+  uint256 tokenId; // Should be 0 for native tokens
+}
```

Comment:
This is true for ERC-20 and ERC-777 tokens, but not for ERC-1155 and ERC-721 tokens.
And since the token created in the bridge is an ERC-1155 we need the `nativeTokenId` to be able to identify the token.

We can change this if we use a "side token factory" that actually creates an ERC-777 token address instead of using the ERC-1155 standard. I believe this will spend more gas to support new tokens but it may be a good approach.

Reply:
In "Future possibilities" section we list "Add support for ERC-721 and ERC-1155 tokens", so I thought they would not be supported for now.

Reply:
Closing as we will only support fungible for now.